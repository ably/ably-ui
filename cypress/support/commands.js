Cypress.Commands.add(
  "compareComponents",
  ({
    reactComponentUrl,
    viewComponentUrl,
    beforeScreenshot = () => {},
    selector,
    componentName = Cypress.mocha.getRunner().suite.ctx.test.parent.title,
  }) => {
    let reactScreenshotPath = null;
    let vwScreenshotPath = null;

    cy.exec(`mkdir -p cypress/screenshots/component-parity/results`, {
      log: false,
    });

    const diffPath = `cypress/screenshots/component-parity/results/${componentName}.png`;

    // If you are wondering why this isn't Promise.all or async/await have a look at
    // https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

    cy.visit(reactComponentUrl);

    beforeScreenshot();

    cy.get(selector)
      .screenshot(`${componentName}.react`, {
        onAfterScreenshot: (_, props) => (reactScreenshotPath = props.path),
      })
      .then(() => {
        cy.visit(viewComponentUrl);
        beforeScreenshot();

        cy.get(selector)
          .screenshot(`${componentName}.vw`, {
            onAfterScreenshot: (_, props) => (vwScreenshotPath = props.path),
          })
          .then(() => {
            cy.exec(
              `cypress/support/diff.js ` +
                `--pathA="${reactScreenshotPath}" ` +
                `--pathB="${vwScreenshotPath}" ` +
                `--target="${diffPath}" ` +
                `--threshold="0.005" `
            ).then((result) => {
              cy.exec(`rm "${reactScreenshotPath}" "${vwScreenshotPath}"`, {
                log: false,
              });

              assert.isTrue(result.stdout === "Yay", "Screenshots match");
            });
          });
      });
  }
);
