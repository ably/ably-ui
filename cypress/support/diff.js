#! /usr/bin/env node

// Used by the compareComponents command

const Blink = require("blink-diff");
const argv = require("yargs").argv;

const diff = new Blink({
  imageAPath: argv.pathA,
  imageBPath: argv.pathB,
  imageOutputPath: argv.target,
  threshold: 0.05,
});

diff.run((error, result) => {
  // We read stdout/stderr - this script is executed interactively by cypress
  if (error) {
    console.error(error);
  } else {
    console.log(diff.hasPassed(result.code) ? "Yay" : "Nay");
  }
});
