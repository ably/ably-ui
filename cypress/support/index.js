// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import "cypress-plugin-snapshots/commands";
import "./commands";

// Cypress has a bug that breaks snapshot tests when running the whole suite
// Follow https://github.com/cypress-io/cypress/issues/3090
// Recommended workaround below:
export const fixSnapshotSpec = (filename) => () => {
  const path = require("path");
  // const relative = filename.substr(1); // removes leading "/"
  const projectRoot = Cypress.config("projectRoot");
  const absolute = path.join(projectRoot, filename);

  Cypress.spec = {
    absolute,
    name: path.basename(filename),
    filename,
  };
};
