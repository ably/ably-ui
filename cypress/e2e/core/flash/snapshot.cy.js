import { fixSnapshotSpec } from "../../../support";

beforeEach(fixSnapshotSpec(__filename));

describe("Flash Snapshot Test", () => {
  beforeEach(() => {
    cy.visit("/components/flash");
    cy.viewport("macbook-16");
  });

  // Flaky
  // it("renders correcty", () => {
  //   cy.get('[data-id="ui-flashes"]');
  //   const flashId = "[data-id='ui-flash']";
  //   cy.get(flashId).eq(0).toMatchSnapshot();
  //   cy.get(flashId).eq(1).toMatchSnapshot();
  //   cy.get(flashId).eq(2).toMatchSnapshot();
  //   cy.get(flashId).eq(3).toMatchSnapshot();
  //   cy.get(flashId).eq(4).toMatchSnapshot();
  // });
});
