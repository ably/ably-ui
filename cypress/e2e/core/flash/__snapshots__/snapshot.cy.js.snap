exports[`Flash Snapshot Test > renders correcty #0`] = `
<div
  class="ui-flash-message ui-grid-px ui-flash-message-enter"
  data-id="ui-flash"
>
  <div
    class="bg-zingy-green p-32 flex align-center rounded shadow-container-subtle"
  >
    <svg
      class="text-cool-black mr-16 self-baseline"
      style="width: 1.5rem; height: 1.5rem;"
    >
      <use xlink:href="#sprite-icon-gui-tick"></use>
    </svg>
    <p class="ui-text-p1 mr-16 text-cool-black">
      Congratulations! You've won the Oscar
    </p>
    <button type="button" class="p-0 ml-auto self-start focus:outline-none">
      <svg
        class="text-cool-black transition-colors"
        style="width: 1.5rem; height: 1.5rem;"
      >
        <use xlink:href="#sprite-icon-gui-close"></use>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Flash Snapshot Test > renders correcty #1`] = `
<div
  class="ui-flash-message ui-grid-px ui-flash-message-enter"
  data-id="ui-flash"
>
  <div
    class="bg-electric-cyan p-32 flex align-center rounded shadow-container-subtle"
  >
    <svg
      class="text-cool-black mr-16 self-baseline"
      style="width: 1.5rem; height: 1.5rem;"
    >
      <use xlink:href="#sprite-icon-gui-ably-badge"></use>
    </svg>
    <p class="ui-text-p1 mr-16 text-cool-black">This is a notice</p>
    <button type="button" class="p-0 ml-auto self-start focus:outline-none">
      <svg
        class="text-cool-black transition-colors"
        style="width: 1.5rem; height: 1.5rem;"
      >
        <use xlink:href="#sprite-icon-gui-close"></use>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Flash Snapshot Test > renders correcty #2`] = `
<div
  class="ui-flash-message ui-grid-px ui-flash-message-enter"
  data-id="ui-flash"
>
  <div
    class="bg-gui-error p-32 flex align-center rounded shadow-container-subtle"
  >
    <svg
      class="text-white mr-16 self-baseline"
      style="width: 1.5rem; height: 1.5rem;"
    >
      <use xlink:href="#sprite-icon-gui-warning"></use>
    </svg>
    <p class="ui-text-p1 mr-16 text-white">This is an error, very bad</p>
    <button type="button" class="p-0 ml-auto self-start focus:outline-none">
      <svg
        class="text-white transition-colors"
        style="width: 1.5rem; height: 1.5rem;"
      >
        <use xlink:href="#sprite-icon-gui-close"></use>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Flash Snapshot Test > renders correcty #3`] = `
<div
  class="ui-flash-message ui-grid-px ui-flash-message-enter"
  data-id="ui-flash"
>
  <div
    class="bg-active-orange p-32 flex align-center rounded shadow-container-subtle"
  >
    <svg
      class="text-white mr-16 self-baseline"
      style="width: 1.5rem; height: 1.5rem;"
    >
      <use xlink:href="#sprite-icon-gui-warning"></use>
    </svg>
    <p class="ui-text-p1 mr-16 text-white">This is an alert</p>
    <button type="button" class="p-0 ml-auto self-start focus:outline-none">
      <svg
        class="text-white transition-colors"
        style="width: 1.5rem; height: 1.5rem;"
      >
        <use xlink:href="#sprite-icon-gui-close"></use>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Flash Snapshot Test > renders correcty #4`] = `
<div
  class="ui-flash-message ui-grid-px ui-flash-message-enter"
  data-id="ui-flash"
>
  <div
    class="bg-electric-cyan p-32 flex align-center rounded shadow-container-subtle"
  >
    <p class="ui-text-p1 mr-16 text-cool-black">
      Some useful information, you are welcome
    </p>
    <button type="button" class="p-0 ml-auto self-start focus:outline-none">
      <svg class=" transition-colors" style="width: 1.5rem; height: 1.5rem;">
        <use xlink:href="#sprite-icon-gui-close"></use>
      </svg>
    </button>
  </div>
</div>
`;
